internal/pkg/etcd/etcd.go

```diff
@@ -103,6 +107,36 @@ func (c *Client) ValidateForUpgrade(ctx context.Context, config config.Provider,
 				return fmt.Errorf("etcd member %d is not started, all members must be running to perform an upgrade", member.ID)
 			}
 		}
+
+		// Member count must not be an even number, lest the uprgade cause a loss of quorum.
+		if len(resp.Members) % 2 == 0 {
+			return fmt.Errorf("etcd member count (%d) is even; upgrade of this node may cause loss of quorum", len(resp.Members))
+		}
+
+		// Quorum must be good before allowing an upgrade of an etcd member.
+		if err = c.ValidateQuorum(ctx); err != nil {
+			return fmt.Errorf("etcd membership quorum validation failed: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// ValidateQuorum performs a KV operation to make certain that quorum is good.
+func (c *Client) ValidateQuorum(ctx context.Context) (err error) {
+
+	// Get a random key. As long as we can get the response without an error, quorum is good.
+	checkCtx, cancel := context.WithTimeout(ctx, QuorumCheckTimeout)
```

Comment:
this changes the timeout for etcd service health check (in service health it's driven by the context passed in), so I think it shouldn't be enforced here (or it should be higher).

Reply:
If the entering context's timeout is shorter, the shorter timeout will still be used.  But a health timeout is different from this quorum check.  If the cluster does not have quorum, this call will block, rather than failing.  Therefore, I believe we _should_ have an additional timeout enforced here, which is separate from the health check timeout.  They are two different things.

Reply:
no, what I mean is: before your changes (you moved this code from the health check), effective timeout was 15s (driven by the health check loop and settings), now timeout is 2s (as it's lower than 15s). I don't know if 2s is too small or not, but it's a change at least for the service health check.

Reply:
if 2s is enough for etcd to commit quorum all the time, please ignore my comment ;)

Reply:
Oh, sorry.  I understand now.  Yeah... I think 15s is too long and also 2s may be too short.  I'll just reset it to 15s for safety.